core/app/models/spree/promotion_code_batch.rb

```diff
@@ -0,0 +1,26 @@
+module Spree
+  class PromotionCodeBatch < ActiveRecord::Base
+    belongs_to :promotion, class_name: "Spree::Promotion"
+
+    validates :number_of_codes, numericality: { greater_than: 0 }
+    validates_presence_of :base_code, :number_of_codes
+
+    def status
+      if error.present?
+        error
+      elsif number_of_codes == number_of_codes_processed
+        "All codes created"
+      else
+        "Processing: #{number_of_codes_processed} / #{number_of_codes}"
+      end
+    end
+
+    def build
```

Comment:
I would prefer a name other than `build`, which will be confusing since ActiveRecord uses that name a lot.

Any of `build_codes`, `create_codes`, `process` would be fine.